File: DiskLockerWindows/disk_locker.cpp
#include <windows.h>
#include <iostream>
#include <vector>
#include <string>

void LockUnlockDisk(char driveLetter, bool lock) {
    std::string drivePath = "\\\\.\\";
    drivePath += driveLetter;
    drivePath += ":";

    HANDLE hDrive = CreateFile(drivePath.c_str(),
                               GENERIC_READ | GENERIC_WRITE,
                               FILE_SHARE_READ | FILE_SHARE_WRITE,
                               NULL, OPEN_EXISTING, 0, NULL);

    if (hDrive == INVALID_HANDLE_VALUE) {
        std::cerr << "Error: Unable to open drive " << driveLetter << std::endl;
        return;
    }

    DWORD bytesReturned;
    if (lock) {
        if (DeviceIoControl(hDrive, FSCTL_LOCK_VOLUME, NULL, 0, NULL, 0, &bytesReturned, NULL)) {
            std::cout << "Drive " << driveLetter << " locked successfully." << std::endl;
        } else {
            std::cerr << "Error: Unable to lock drive " << driveLetter << std::endl;
        }
    } else {
        if (DeviceIoControl(hDrive, FSCTL_UNLOCK_VOLUME, NULL, 0, NULL, 0, &bytesReturned, NULL)) {
            std::cout << "Drive " << driveLetter << " unlocked successfully." << std::endl;
        } else {
            std::cerr << "Error: Unable to unlock drive " << driveLetter << std::endl;
        }
    }

    CloseHandle(hDrive);
}

void DisplayDisks() {
    DWORD drives = GetLogicalDrives();
    std::cout << "Available drives:" << std::endl;

    for (int i = 0; i < 26; ++i) {
        if (drives & (1 << i)) {
            std::cout << char('A' + i) << ":\\ ";
        }
    }
    std::cout << std::endl;
}

int main() {
    int choice;
    char driveLetter;
    std::vector<char> selectedDisks;
    bool lockAll = false, unlockAll = false;

    std::cout << "1. Lock all disks\n2. Unlock all disks\n3. Select specific disks to lock\n4. Select specific disks to unlock" << std::endl;
    std::cout << "Enter your choice: ";
    std::cin >> choice;

    switch (choice) {
    case 1:
        lockAll = true;
        break;
    case 2:
        unlockAll = true;
        break;
    case 3:
        std::cout << "Select disks to lock (enter drive letters, finish with 0):" << std::endl;
        DisplayDisks();
        while (true) {
            std::cin >> driveLetter;
            if (driveLetter == '0') break;
            selectedDisks.push_back(driveLetter);
        }
        break;
    case 4:
        std::cout << "Select disks to unlock (enter drive letters, finish with 0):" << std::endl;
        DisplayDisks();
        while (true) {
            std::cin >> driveLetter;
            if (driveLetter == '0') break;
            selectedDisks.push_back(driveLetter);
        }
        break;
    default:
        std::cerr << "Invalid choice" << std::endl;
        return 1;
    }

    if (lockAll) {
        DWORD drives = GetLogicalDrives();
        for (int i = 0; i < 26; ++i) {
            if (drives & (1 << i)) {
                LockUnlockDisk(char('A' + i), true);
            }
        }
    } else if (unlockAll) {
        DWORD drives = GetLogicalDrives();
        for (int i = 0; i < 26; ++i) {
            if (drives & (1 << i)) {
                LockUnlockDisk(char('A' + i), false);
            }
        }
    } else {
        for (char disk : selectedDisks) {
            LockUnlockDisk(disk, choice == 3);
        }
    }

    return 0;
}

File: DiskLockerWindows/DiskLocker.java
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class DiskLocker {

    public static void lockUnlockDisk(char driveLetter, boolean lock) {
        String drivePath = driveLetter + ":\\";
        File disk = new File(drivePath);

        if (!disk.exists()) {
            System.out.println("Error: Unable to access drive " + driveLetter);
            return;
        }

        // Здесь нужно добавить код для блокировки/разблокировки диска,
        // используя соответствующие Java библиотеки или нативные вызовы.
        // В данном примере просто выводим сообщение.
        if (lock) {
            System.out.println("Drive " + driveLetter + " locked successfully.");
        } else {
            System.out.println("Drive " + driveLetter + " unlocked successfully.");
        }
    }

    public static List<Character> displayDisks() {
        List<Character> disks = new ArrayList<>();
        for (char drive = 'A'; drive <= 'Z'; drive++) {
            String drivePath = drive + ":\\";
            if (new File(drivePath).exists()) {
                disks.add(drive);
            }
        }
        return disks;
    }

    public static void main(String[] args) {
        List<Character> disks = displayDisks();
        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("Available drives: " + disks);
            System.out.println("Enter drive letter to lock/unlock (or 'exit' to quit): ");
            String input = scanner.nextLine();

            if (input.equalsIgnoreCase("exit")) {
                break;
            }

            if (input.length() == 1 && disks.contains(input.charAt(0))) {
                System.out.println("Type 'lock' to lock or 'unlock' to unlock: ");
                String action = scanner.nextLine();

                if (action.equalsIgnoreCase("lock")) {
                    lockUnlockDisk(input.charAt(0), true);
                } else if (action.equalsIgnoreCase("unlock")) {
                    lockUnlockDisk(input.charAt(0), false);
                } else {
                    System.out.println("Invalid action.");
                }
            } else {
                System.out.println("Invalid drive letter.");
            }
        }

        scanner.close();
    }
}
